### PR TITLE
Chore: Delete old version/Calculate old version in AYON Loader

### PR DESCRIPTION
**Pull request by moonyuet**
## Changelog Description
This PR is to fix the issue of delete old version/calculate old version do not work in AYON Loader

## Additional info
Ayon server / loader would still list the old versions even though they are deleted.

## Testing notes:
1. Launch AYON
2. Right Click AYON icon to access Loader
3. Right Click any instance
4. Delete old version
image

### CODE REVIEW

**Comment by BigRoy** [comment]
Any chance you can "squash" your commits or rebase it in some way? Github changes really doesn't seem to like 'showing the changes' nicely with the file being deleted in-between in the commits?

**Review by LiborBatek — APPROVED** [review]
Did few tests with different ```Delete``` options and all working as expected...

image

I have also checked the publish folders if all correct after triggering the action. All good.

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 1]
```diff
@@ -1,501 +1,478 @@
...
+        msgBox.setWindowFlags(
+            msgBox.windowFlags() | QtCore.Qt.FramelessWindowHint
+        )
+        msgBox.exec_()
+
+    def get_data(self, context, versions_count):
+        product_entity = context["product"]
+        folder_entity = context["folder"]
+        project_name = context["project"]["name"]
+        anatomy = Anatomy(project_name)
```
Comment (iLLiCiTiT) [inline client/ayon_core/plugins/load/delete_old_versions.py:199]:
```suggestion
        anatomy = Anatomy(project_name, project_entity=context["project"])
```

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 2]
```diff
@@ -1,501 +1,478 @@
...
+        try:
+            session.commit()
+
+        except Exception:
+            msg = (
+                "Could not set `is_published` attribute to `False`"
+                " for selected AssetVersions."
+            )
+            self.log.error(msg)
+            self.message(msg)
```
Comment (BigRoy) [inline client/ayon_core/plugins/load/delete_old_versions.py:417]: Should this PR include Ftrack addon specific testing? Which may be hard to do as long as the Ftrack addon may not be fully up-to-date with the refactored `ayon-core`?

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 3]
```diff
@@ -1,501 +1,478 @@
...
+        # Pop dirs from both dictionaries
+        for dir_id in dir_ids_to_pop:
+            dir_paths.pop(dir_id)
+            paths = file_paths_by_dir.pop(dir_id)
+            # TODO report of missing directories?
+            paths_msg = ", ".join([
+                "'{}'".format(path[0].replace("\\", "/")) for path in paths
+            ])
+            self.log.debug((
+                "Folder does not exist. Deleting it's files skipped: {}"
```
Comment (BigRoy) [inline client/ayon_core/plugins/load/delete_old_versions.py:315]:
```suggestion
                "Folder does not exist. Deleting its files skipped: {}"
```

File: client/ayon_core/plugins/load/delete_old_versions.py (hunk 2 above)
Comment (moonyuet) [reply client/ayon_core/plugins/load/delete_old_versions.py:417]: > Should this PR include Ftrack addon specific testing? Which may be hard to do as long as the Ftrack addon may not be fully up-to-date with the refactored `ayon-core`?
@iLLiCiTiT

Comment (iLLiCiTiT) [reply client/ayon_core/plugins/load/delete_old_versions.py:417]: Ha, I didn't know it's there is ftrack removal. Should be removed.

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 4]
```diff
@@ -1,501 +1,409 @@
...
+        msgBox.exec_()
+
+    def get_data(self, context, versions_count):
+        product_entity = context["product"]
+        folder_entity = context["folder"]
+        project_name = context["project"]["name"]
+        anatomy = Anatomy(project_name, project_entity=context["project"])
+
+        versions = list(ayon_api.get_versions(
+            project_name, product_ids=[product_entity["id"]]
```
Comment (iLLiCiTiT) [inline client/ayon_core/plugins/load/delete_old_versions.py:201]: Make sure all versions active/inactive are fetched, and skip hero version.
```suggestion
            project_name,
            product_ids=[product_entity["id"]],
            active=None,
            hero=False,
```

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 5]
```diff
@@ -1,501 +1,409 @@
...
+        if remove_publish_folder:
+            size = self.delete_whole_dir_paths(data["dir_paths"].values())
+        else:
+            size = self.delete_only_repre_files(
+                data["dir_paths"], data["file_paths_by_dir"]
+            )
+
+        for version in data["versions"]:
+            orig_version_tags = version["data"].get("tags") or []
+            version_tags = [tag for tag in orig_version_tags]
```
Comment (iLLiCiTiT) [inline client/ayon_core/plugins/load/delete_old_versions.py:341]:
```suggestion
            orig_version_tags = version["tags"]
            version_tags = list(orig_version_tags)
```

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 6]
```diff
@@ -1,501 +1,409 @@
...
+            if "deleted" not in version_tags:
+                version_tags.append("deleted")
+
+            if version_tags == orig_version_tags:
+                continue
+
```
Comment (iLLiCiTiT) [inline client/ayon_core/plugins/load/delete_old_versions.py:347]:
```suggestion
```

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 7]
```diff
@@ -1,501 +1,409 @@
...
-#             version_tags = [tag for tag in orig_version_tags]
-#             if "deleted" not in version_tags:
-#                 version_tags.append("deleted")
-#
-#             if version_tags == orig_version_tags:
-#                 continue
-#
-#             update_query = {"id": version["id"]}
-#             update_data = {"$set": {"data.tags": version_tags}}
-#             mongo_changes_bulk.append(UpdateOne(update_query, update_data))
```
Comment (iLLiCiTiT) [inline client/ayon_core/plugins/load/delete_old_versions.py:353]: Where is conversion of this logic?

Comment (moonyuet) [reply client/ayon_core/plugins/load/delete_old_versions.py:353]: are we still using the mongo or we can use some postsql?

File: client/ayon_core/plugins/load/delete_old_versions.py [hunk 8]
```diff
@@ -1,501 +1,420 @@
...
+            versions.remove(version)
+
+        # Update versions_by_parent without filtered versions
+        versions_by_parent = collections.defaultdict(list)
+        for ent in versions:
+            versions_by_parent[ent["productId"]].append(ent)
+
+        # Filter already deleted versions
+        for version in versions:
+            if version["active"] or "deleted" in version["tags"]:
```
Comment (iLLiCiTiT) [inline client/ayon_core/plugins/load/delete_old_versions.py:239]: My fault, this probably should be only check for tags...
```suggestion
            if "deleted" in version["tags"]:
```

**Review by LiborBatek — CHANGES_REQUESTED** [review]
Do not work atm and gives following error msg:

```
ERROR:ProductLoader:Failed to delete versions.
Traceback (most recent call last):
  File "C:\Work\REPO\ayon-core\client\ayon_core\plugins\load\delete_old_versions.py", line 375, in load
    data = self.get_data(context, versions_to_keep)
  File "C:\Work\REPO\ayon-core\client\ayon_core\plugins\load\delete_old_versions.py", line 239, in get_data
    if "deleted" in version["tags"]:
KeyError: 'tags'
```

**Comment by LiborBatek** [comment]
When performing ```Calculate old versions``` Im getting this error now:

```
ERROR:ProductLoader:Failed to delete versions.
Traceback (most recent call last):
  File "C:\Work\REPO\ayon-core\client\ayon_core\plugins\load\delete_old_versions.py", line 382, in load
    size += self.main(project_name, data, remove_publish_folder)
TypeError: main() takes 3 positional arguments but 4 were given
```

**Comment by LiborBatek** [comment]
When performing ```Delete Old Versions``` and set it to keep just one version...even tho it doesnt give any error message it doesnt remove any versions ??

image

**Review by LiborBatek — CHANGES_REQUESTED** [review]
Did another test round this time ```Delete Old Versions (keep 5 workfiles)``` 

Screenshot 2024-04-15 143821

Note my file explorer on the left side of the scrn and its workfiles 14 versions present.

And after triggering the ```Delete``` function

Screenshot 2024-04-15 144154

It seems like those 5 last versions was kept ok (as I was opting for) and the rest being somewhat corrupted aka not removed properly but just partially.

**Review by LiborBatek — APPROVED** [review]
Now it correctly removes the versions you are after. LGTM

**Comment by moonyuet** [comment]
@iLLiCiTiT Is it okay to be merged now?

**Pull request merged**